compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp

```diff
@@ -0,0 +1,386 @@
+/*
+ * Copyright (c) 2023 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/RemoveUnnecessaryTransposeNetPass.h"
+
+#include <luci/IR/CircleNodes.h>
+#include <luci/Profile/CircleNodeOrigin.h>
+
+#include <vector>
+
+namespace
+{
+
+class TaggedShapeAnalyzer final
+{
+public:
+  /**
+   * @brief check 'Transpose-Reshape-Transpose' can be replaced by one 'Reshape'.
+   *
+   * NOTE:
+   *  - It tracks How each dimension of shape is transformed/moved by using 'tags'.
+   *    Tag records the initial order of each dimension.
+   *
+   *  - Tag is only attached for dimensions with non-1 value.
+   *
+   *  - For details, please refer analyze_(transpose, reshape) comments.
+   */
+  bool canRemoveTransposes(const luci::CircleTranspose *f_tr, const luci::CircleReshape *m_rs,
+                           const luci::CircleTranspose *b_tr);
+
+private:
+  void init_shape_with_tag(const luci::CircleNode *in_tensor);
+  bool analyze_transpose(const luci::CircleTranspose *transpose_node);
+  bool analyze_reshape(const luci::CircleReshape *reshape_node);
+  bool verify_tag() const;
+
+  struct Dim final
+  {
+    uint32_t value;
+    std::vector<uint8_t> tags;
+  };
+  using Shape = std::vector<Dim>;
+
+  const uint8_t _start_tag = 0;
```

Comment:
Q) why use `const uint8_t` instead if `#define START_TAG 0` ?

Reply:
I thought the start-value of a tag is logically up to `TaggedShapeAnalyzer`. 
So I hope, the relationship between`TaggedShapeAnalyzer` and `_tag_start` is shown, somehow. 
So, I put it inside the class. 

+) I also thought that `#define` statement isn't bounded with namespace. So I thought I had to avoid using unnecessary `#define`. 

I'm open to others' opinions about this. 
If you have another idea, please let me know.

Reply:
`_start_tag ` gives me as a variable when reading this code, not constant value for index initialization.
Maybe `const uint8_t START_TAG = 0` can let readers that this is fixed constant, not a variable.